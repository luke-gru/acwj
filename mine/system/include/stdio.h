#ifndef _STDIO_H_
#define _STDIO_H_

#include <stddef.h>
#include <stdarg.h>

#define EOF (-1)

// This FILE definition will do for now
typedef char *FILE;
extern FILE *stderr;
extern FILE *stdout;

FILE *fopen(char *pathname, char *mode);
size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
size_t fwrite(void *ptr, size_t size, size_t nmemb, FILE *stream);
int fclose(FILE *stream);
int printf(char *format);
int fprintf(FILE *stream, char *format);
size_t getline(char **lineptr, size_t *n, FILE *stream);

FILE *popen(const char *command, const char *type);
int pclose(FILE *stream);

int vfprintf(FILE *stream, const char *format, va_list ap);

#endif	// _STDIO_H_
