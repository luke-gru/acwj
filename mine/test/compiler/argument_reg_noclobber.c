#include <stdio.h>

int mycall(int a, int b, int c, int d, int e) {
  return (1);
}

int main(void) {
  long i;
  i = 10;
  printf("%ld\n", i);
  /**
   * mycall() shouldn't clobber argument register 4 for printf() (rcx)
   * In order to do this, for a call to a function we make a first-pass thru
   * the argument list (forwards), and associate
   * them with an argument position. Those values are pushed on the stack.
   * Then when walking backwards to generate the cgloadargs, we pop the value
   * off the stack if it's a function call
   *
   * Try this first:
   * Brute force: for every function call cgloadarg, we push all argument
   * values to the stack before the call and pop them off after the call.
   */
  printf("%ld, %d, %ld\n", i, mycall(1, 2, 3, 4, 5), i);
  return (0);
}
