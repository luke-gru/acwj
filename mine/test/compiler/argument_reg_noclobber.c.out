10
10, 1, 10
